Self hosted additions to documentation (#386)
MicrosoftDocs/vsonline: vs-online/how-to/self-hosting-cli.md

### macOS

Homebrew is the easiest way to manage your CLI install. It provides convenient ways to install, update, and uninstall. If you don't have homebrew available on your system, [install homebrew](https://docs.brew.sh/Installation.html) before continuing.

You can install the CLI by updating your brew repository information, and then running the install command:

```bash
brew update && install microsoft/vsonline/vso
```

### Linux

If you are running a distribution that comes with `apt`, such as Ubuntu or Debian, run the following commands to install the `vso` package:

```bash
sudo add-apt-repository https://packages.microsoft.com/ubuntu/18.04/prod/
sudo apt-get update
sudo apt-get install vso
```

For other Linux distributions, run the following script by using curl and pipe directly to bash, or download the script to a file and inspect it before running.

```bash
sudo wget -O ~/vso-script https://aka.ms/install-vso-linux && sudo tr -d '\r' < ~/vso-script  | bash
```

## Register your machine

>[!NOTE]
> Ensure you have created a VS Online plan before starting the registration process. Learn how to [create a plan](browser.md#create-a-plan) in our browser how-to guide.

> Once you've successfully installed the CLI, start the interactive registration process by running `vso start`. Be sure you're running the process as the user you'd like to use when accessing the environment from other machines after registration.

Reviewer: Thanks for the note about running the process as the user. It might be helpful to explicitly add something to the effect of "We do not recommend running with sudo"

As part of registration, you will be asked to sign-in with your VS Online account and provide the following information:

> - **Plan Name**: The existing plan name this environment should be registered to

Reviewer: Right now, we just prompt the user to select an existing plan ID. When they invoke `vso start` they can supply the argument `--plan-name` to filter out any of their plans that do not include that planId.

Later on, we will have an optional plan creation flow. As part of that flow, we _will_ prompt for plan name.

> - **Workspace Path**: The workspace path for the environment

Reviewer: We do not ask for this - if it is supplied in the arguments, we use that. Otherwise we use the current working directory.

Maybe worth noting somewhere what this means and it's security implications: it's just a landing spot when you connect, but you'll be able to switch folders, etc.

Author: Adding in. Won't show here due to some refactoring of the files but I've taken this into account.

- **Name**: What you'd like to name this environment
- **Register as a service or run as a process**: Choose how you would like the VS Online agent to run
  - **Register a service:** Allows the agent to be registered as a system service and persists after environment restarts.
      > [!NOTE]

>       > You must have administrator permissions to be able to select this option. You will be prompted to provide an admin password to complete registration.

Reviewer: Some clarification:

On Windows: you must also run `vso start` from a terminal/command line that is running as an admin user. You will be prompted for the password of that user specifically.
On Mac, you may be prompted for a sudoers password (unless another sudo command has recently been successfully authenticated/authorized)
On Linux, you will not be prompted for a password, but you will be prompted if you want to run `loginctl enable-linger` to keep the agent running even when logged off.

Author: Added to instructions for CLI registration process.

---
author: abuchholtzau
ms.author: allisb
ms.service: visual-studio-online

> title: Register a self-hosted, headless VS Online environment

Reviewer 2: Swapping order based on the header below
```suggestion
title: Register a headless, self-hosted VS Online environment
```

ms.topic: overview
ms.date: 04/06/2020
---

# Register a headless, self-hosted VS Online environment

> You can host VS Online on your own environment using the headless CLI. This article describes how to register and connect to a headless VS Online server.

Reviewer 2: ```suggestion
You can host a VS Online environment on your own hardware using the headless CLI. This article describes how to register and connect to a headless VS Online server.
```

## Install VS Online for self-hosting

VS Online CLI tools can be used on [Windows](~/reference/vsonline-cli.md#windows), [macOS](~/reference/vsonline-cli.md#macos), and [Linux](~/reference/vsonline-cli.md#linux) environments. For installation instructions, see [VS Online CLI Reference](~/reference/vsonline-cli.md#installation).

## Register your machine

You must create a VS Online plan before starting the registration process. For more information on creating a plan, see [create a plan](browser.md#create-a-plan) in our browser how-to guide.

Once you've successfully installed the CLI, start the interactive registration process by running the [`vso start`](~/reference/vsonline-cli.md#start-vs-online) command.

> [!NOTE]
> Be sure you're running the process as the user you'd like to use when accessing the environment from other machines after registration.

### Register as a service

> [!NOTE]
> You must have administrator permissions to be able to select this option. For macOS and Linux, we do not recommend running with sudo.

> As part of registration, you will be asked to sign-in with your VS Online account and provide the existing plan name this environment should be registered to.

Reviewer 2: ```suggestion
As part of registration, you will be asked to sign-in with your VS Online account and provide the existing plan this environment should be registered to.
```

#### Windows

Register the self-hosted agent as a service by running the following command from a terminal/command line that is running as an admin user.

```bash
vso start --service
```

> You will be prompted for the password of that user specifically.

Reviewer: Maybe "You will be prompted for your Windows login password."

I anticipate that users will be wondering what we do with the user password. We could add that to our docs in a separate PR.

#### macOS

Register the self-hosted agent as a service by running:

```bash
vso start --service
```

You may be prompted for a sudoers password (unless another sudo command has recently been successfully authenticated/authorized).

#### Linux

>[!NOTE]
> `systemd` must be installed to run as a service for Linux.

Register the self-hosted agent as a service by running:

```bash
vso start --service
```

You will be prompted to confirm if you want to run `loginctl enable-linger` to keep the agent running even when logged off.

### Run as a process

> Once you've successfully installed the CLI, register the self-hosted agent as a service by running:

Reviewer 2: ```suggestion
> [!NOTE]
> When running as a temporary process, the agent will stop as soon as the process exits. This will make the environment switch to an Unavailable stat,  and you will not be able to connect to the environment until the agent process is restarted. If you'd like the agent to be installed long-term, consider running as a service instead.

Once you've successfully installed the CLI, run the self-hosted agent as a temporary process by running:
```

```bash
vso start
```

> As part of registration, you will be asked to sign-in with your VS Online account and provide the existing plan name this environment should be registered to.

Reviewer: nit: probably change "plan name" to just "plan" (technically plan names do not uniquely identify a plan)
